### PR TITLE
player code updates/refactoring

### PR DESCRIPTION
**Pull request by nukeulater**
- re-added pre-spawn player character type validation

### CODE REVIEW

**Comment by nukeulater** [comment]
@Berthalamew lmk if everything looks fine to you

File: xlive/Blam/Engine/game/players.cpp [hunk 1]
```diff
@@ -306,11 +311,58 @@ void __cdecl player_validate_configuration(datum player_index, s_player_properti
...
     // Change the validation for player_appearance_valid to use the updated k_player_character_type_count constant
     WriteValue<BYTE>(Memory::GetAddress(0x54fb3, 0x5D4AB), k_player_character_type_count);
 
     // Replace the player profile validation function with our own
     PatchCall(Memory::GetAddress(0x5509E, 0x5D596), player_validate_configuration);
+
+    // Validate the player character type pre-spawn
+    p_player_representation_get = Memory::GetAddress<void*>(0x53895, 0x5BD8D);
+    PatchCall(Memory::GetAddress(0x559F9, 0x5DEF1), player_representation_get_to_cdecl);
+    PatchCall(Memory::GetAddress(0x53969, 0x5BE61), player_representation_get_to_cdecl);
```
Comment (Berthalamew) [inline xlive/Blam/Engine/game/players.cpp:388]: These functions shouldn't be hooked for verification.
Detour the player_spawn function instead and add a call to player_configuration_validate_character_type.

File: xlive/Blam/Engine/game/players.cpp [hunk 2]
```diff
@@ -97,7 +97,7 @@ datum s_player::get_unit_index(datum player_index)
 	return get(player_index)->unit_index;
 }
 
-uint8* s_player::get_player_unit(datum player_index) 
+uint8* s_player::get_player_unit_data(datum player_index)
```
Comment (Berthalamew) [inline xlive/Blam/Engine/game/players.cpp]: this should return unit_datum*

**Review by Berthalamew — CHANGES_REQUESTED** [review]

File: xlive/Blam/Engine/game/players.cpp (hunk 1 above)
Comment (nukeulater) [reply xlive/Blam/Engine/game/players.cpp:388]: uh, this is what this does? it hooks a function within player spawn to validate the character type pre-spawn

Comment (Berthalamew) [reply xlive/Blam/Engine/game/players.cpp:388]: I meant detouring player_spawn directly and executing the validation before player_spawn.
The function you're hooking currently is used twice during one of the spawn processes which is unnecessary.

Comment (nukeulater) [reply xlive/Blam/Engine/game/players.cpp:388]: I don't see a problem with it happening twice

**Pull request merged**